### DISCUSSION TITLE
How can `pool_slots` work effectively across multiple worker nodes?

### DISCUSSION THREAD

**Discussion by Lyalpha** [apache/airflow#26133] [Q&A] [answered]
(I strongly suspect being able to do this is all outside the scope airflow, but would be interested to hear otherwise, or if there are plans to implement something like this - by building on the `pool_slots` feature.)

---
#### Problem
1. An airflow instance running with `default_pool_task_slot_count = 10`.

2 .Two Celery worker nodes processing these, each with `worker_concurrency = 5`.

3 .Normal, light-weight tasks are processed fine. 

4 . There is a heavy-weight task, which takes up an entire worker nodes memory to run - we can set this task to have `pool_slots=5` so we might expect only one instance of the task can run on a given node. However, although at most two instances of this task will be scheduled by airflow, they can conceivably both end up on the same worker, since each still occupies only one "concurrency" slot of the worker.

---

To solve this I have the partial work around where each worker `X` has its own `queue_X` and associated `pool_X`, but two main drawbacks:
* One has to set the pool/queue for the task when authoring the DAG. There is no way - afaik - to assign it randomly to a pool/queue pair (or, even better, a not-full pair)
* I have to set aside the possibility that heavy-weight tasks may be running concurrently with a large number of light-weight tasks, since they have no knowledge of each other, meaning an under-utilisation of resources.

If I could define a single default pool size, and tasks could be assigned to occupy a number of `concurrency_slots` instead, it would overcome these.

**Comment by potiuk** [comment, marked as answer]
I don't think it is easy to be done with Celery because Celery is not really even foreseen in handling resource allocation (and neither is Airflow). But I think what might work is going K8S simply employing CeleryKubernetesExecutor and assign your heavy tasks to "kubernetes" pool. Then you could define your pod template to define all kinds of directives (for example anti-affinity to not allow two similar task pods to be executed at the same node).

**Reply by Lyalpha** [reply]
OK thanks, suspected so. I don't think I have the resources to start looking into using Kubernetes.
